enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,486 @@
+enabled, the new admission hook will mutate all incoming pods, adding requests/limits for the new
+extended resources. This means that all nodes in the cluster must have a Performance Profile that
+includes shared CPUs - without that, no pods would be scheduled on that node. Note that the isolated
+CPUs partition would be optional and would only be required on nodes that were going to run
+pods that had containers with whole CPU requests/limits.
+
+### Risks and Mitigations
+
+Nothing yet...
+
+### Drawbacks
+
+The changes proposed to kubelet will require us to carry patches to the upstream version. The idea of
+splitting the CPUs used by kubelet into a pre-defined shared and guaranteed pool might be considered
+upstream, but the use of the extended resources to track these is unlikely to be accepted. The
+alternative of adding a new first class resource (e.g. cpu-guaranteed) instead of extended resources
+would have huge impacts to existing code and is unlikely to be accepted.
+
+Another drawback is that the number of shared and guaranteed CPUs can only be changed with a reboot. However,
+there is no alternative that avoids a reboot since much of the required kernel tuning is static and cannot
+be changed at runtime.
+
+## Design Details
+
+### Open Questions [optional]
+
+#### Enabling Feature
+
+How will this be enabled at the openshift apiserver level? It looks like you can create an admission hook 
+that is disabled by default and then enabled through a config option on the API server - could that mechanism
+be used?
+
+#### Scheduler Awareness
+
+In the current proposal, the scheduler is aware of the new `openshift.io/shared-cpus` and
+`openshift.io/guaranteed-cpus` resources and will ensure that a pod is only scheduled on a node with enough
+of those resources available. However, this does not account for cases where the kubelet has a more
+restrictive Topology Manager Policy (e.g. `single-numa-node` policy). In that case, it is possible that a pod
+could be scheduled on a node that had enough total guaranteed CPUs (for example), but not enough guaranteed
+CPUs on a single NUMA node. This would result in kubelet rejecting the pod with a Topology Affinity error.
+
+One way to address this would be through the Topology Aware Scheduler and RTE/NFD - adding scheduler
+awareness of how many shared/guaranteed CPUs are available on each NUMA node. This will be explored further
+in the Topology Aware Scheduler alternative below.
+
+Note: In practice, this would only be an issue in multi-node deployments where there are pods that could
+run on a selection of nodes. In a single node deployment, the user will configure the number of shared and 
+guaranteed CPUs to match the workloads that they are planning on running on the node. The user would be 
+aware of the NUMA restrictions imposed by using the `single-numa-node` policy and would ensure their 
+configuration matched.
```

Comment:
It turns out that scheduler awareness isn’t an issue (or at least isn’t any worse than it is without this enhancement). Prior to this enhancement, using the single-numa-node policy in kubelet’s Topology Manager can result in a Topology Affinity error when, for example, a container requests two guaranteed CPUs but the only two CPUs available are on two different NUMA nodes. The scenario is the same with the enhancement - the Topology Affinity error can occur if there are only two openshift.io/guaranteed-cpus available but they are on two different NUMA nodes.

If the user wants to avoid this scenario, the solution will still be the Topology Aware Scheduler (TAS). This should work with a very small change to the kubelet - when reporting the cpu resources available per NUMA node through the PodResources API, instead of reporting all available CPUs, when this shared/guaranteed CPU feature is enabled, we would only report the available guaranteed CPUs. This will ensure that the TAS will only schedule pods to nodes where there are enough guaranteed CPUs available on the same NUMA node (when the single-numa-node policy is being used).

Finally, after looking at the TAS/NFD/RTE, it isn’t going to be possible to use it as a solution to enable the shared/guaranteed CPU partition proposed in this enhancement. The main reason being that TAS/NFD/RTE can’t be deployed on some SNO configurations (e.g. vRAN DU) due to resource constraints. We need a solution that can be used in those configurations. A secondary reason is that the shared CPUs are not NUMA node specific - containers using shared CPUs float across all the shared CPUs - they are not allocated to specific containers so this doesn’t map well to the NRT and the role of the TAS. Tracking the shared/guaranteed CPUs as extended resources is much more intuitive and easily visible to the user (i.e. visible in “oc describe node” vs. the NRT CR).

Reply:
> instead of reporting all available CPUs, when this shared/guaranteed CPU feature is enabled, we would only report the available guaranteed CPUs.

Didn't think it through but at glance that could work.
And still TAS would have to learn about the shared-cpus quantity, but that can be done by reading the information from Kubelet. 

> The main reason being that TAS/NFD/RTE can’t be deployed on some SNO configurations (e.g. vRAN DU) due to resource constraints.

That's too bad.

Reply:
Updated the enhancement based on the comments above.